File: src/lib/permissions.ts
// TS mirror of `agent/src-tauri/src/permissions.rs` (#623). Keeps the
// Tauri IPC command names + the `CapturePermissions` / `PermStatus`
// shape in one place so the record-page pre-flight and the onboarding
// permissions slide call the same wrappers.
//
// macOS gates native capture (cpal mic + ScreenCaptureKit loopback)
// behind TCC grants. Off macOS every status is `not_applicable` and
// the request commands no-op — the contract stays uniform so callers
// never need a per-OS branch beyond "is this status `denied`".

import { invoke } from "@tauri-apps/api/core";

/** Per-permission grant state. Wire strings match the Rust
 *  `#[serde(rename_all = "snake_case")]` enum. */
export type PermStatus =
  | "granted"
  | "denied"
  | "undetermined"
  | "not_applicable";

/** Combined snapshot from `check_capture_permissions`. */
export type CapturePermissions = {
  microphone: PermStatus;
  screen_recording: PermStatus;
};

/** Which macOS Privacy & Security pane `openPrivacySettings` targets. */
export type PrivacyPane = "microphone" | "screen";

/** Read the live grant state for mic + screen-recording capture.
 *  Cheap status read (never prompts) so it's safe on the hot Start
 *  path. */
export function checkCapturePermissions(): Promise<CapturePermissions> {
  return invoke<CapturePermissions>("check_capture_permissions");
}

/** Fire the OS mic-permission prompt (macOS) and resolve with the
 *  resulting status. `not_applicable` off macOS. */
export function requestMicPermission(): Promise<PermStatus> {
  return invoke<PermStatus>("request_mic_permission");
}

/** Prompt for screen-recording access (macOS) and resolve true when
 *  granted. Always true off macOS. */
export function requestScreenCaptureAccess(): Promise<boolean> {
  return invoke<boolean>("request_screen_capture_access");
}

/** Open the relevant macOS Privacy & Security pane so the user can
 *  flip a denied grant. Non-fatal: opener failures reject, callers
 *  should swallow.
 *
 *  NOTE: the Rust command is registered but its body is still a stub
 *  (TODO #623 S-4) — this currently resolves without opening anything.
 *  Do not wire an "Open System Settings" affordance to it until S-4
 *  lands the deep-link body. */
export function openPrivacySettings(pane: PrivacyPane): Promise<void> {
  return invoke("open_privacy_settings", { pane });
}

File: src-tauri/src/permissions.rs
// macOS capture-permission pre-flight (#623).
//
// Capture is native — `cpal` for the mic, ScreenCaptureKit for the
// system loopback — NOT `getUserMedia`, so the OS gates capture behind
// TCC (Transparency, Consent, and Control) grants rather than a
// webview permission prompt. A *denied* mic permission would otherwise
// surface as a raw `cpal::BuildStreamError` string at the recording
// aha-moment; this module lets the frontend read the grant state
// up-front and short-circuit the already-denied case with an
// actionable message instead of a cryptic error.
//
// This is the seam only — the pre-flight gating on the Start path and
// the onboarding permissions slide that consume it are separate work.
//
// Platform split:
//   - macOS: query AVFoundation (mic) + CoreGraphics (screen recording)
//     through the existing `swift-bridge` FFI seam in
//     `macos_loopback.rs` — no new crate.
//   - Linux / Windows: PipeWire / WASAPI do not gate capture behind a
//     TCC-style per-app grant, so both permissions report
//     `not_applicable` and the request helpers are no-ops. The macOS
//     arm is the point of this module; these arms keep the IPC
//     contract uniform across builds.

use serde::Serialize;

/// Per-permission grant state. Serialised snake_case so the TS mirror
/// in `agent/src/lib/permissions.ts` can match on the wire strings
/// directly.
///
/// Off macOS only `NotApplicable` is ever constructed (PipeWire/WASAPI
/// don't gate capture behind a TCC grant), so the granted/denied/
/// undetermined variants are legitimately dead on those builds — they
/// stay in the enum because they're the cross-platform wire contract.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(not(target_os = "macos"), allow(dead_code))]
pub enum PermStatus {
    /// Capture is allowed.
    Granted,
    /// The user explicitly denied (or the grant is restricted by MDM).
    Denied,
    /// No decision yet — the OS will prompt on first capture.
    Undetermined,
    /// Platform doesn't gate this capture behind a TCC-style grant
    /// (Linux / Windows), so there is nothing to check or request.
    NotApplicable,
}

/// Combined snapshot returned by `check_capture_permissions`.
#[derive(Debug, Clone, Copy, Serialize)]
pub struct CapturePermissions {
    pub microphone: PermStatus,
    pub screen_recording: PermStatus,
}

#[cfg(target_os = "macos")]
mod imp {
    use super::{CapturePermissions, PermStatus};
    use crate::macos_loopback::ffi;

    // AVFoundation `AVAuthorizationStatus` raw values
    // (notDetermined=0, restricted=1, denied=2, authorized=3) — kept
    // here rather than in Swift so the mapping is auditable on the
    // Rust side. The Swift shim returns the raw `Int`.
    fn map_av_status(raw: i32) -> PermStatus {
        match raw {
            3 => PermStatus::Granted,
            0 => PermStatus::Undetermined,
            // 1 (restricted) + 2 (denied) both leave the user blocked
            // with the same remedy (System Settings), so collapse them.
            _ => PermStatus::Denied,
        }
    }

    // ScreenCaptureKit preflight is a bool: granted (true) or not
    // (false). There is no "undetermined" distinct from "denied" at
    // the CoreGraphics preflight layer, so a `false` maps to denied —
    // screen-recording denial is a soft, dismissible degrade
    // (mic-only), so the granted/denied split is sufficient.
    fn map_screen_status(granted: bool) -> PermStatus {
        if granted {
            PermStatus::Granted
        } else {
            PermStatus::Denied
        }
    }

    pub fn check() -> CapturePermissions {
        CapturePermissions {
            microphone: map_av_status(ffi::micAuthStatus()),
            screen_recording: map_screen_status(ffi::screenCaptureAuthStatus()),
        }
    }

    pub fn request_mic() -> PermStatus {
        // Fires the AVFoundation request prompt (blocks until the user
        // answers), then re-reads the authoritative status.
        ffi::requestMicAccess();
        map_av_status(ffi::micAuthStatus())
    }

    pub fn request_screen() -> bool {
        // `CGRequestScreenCaptureAccess()` prompts (or no-ops if a
        // decision exists) and returns the resulting grant.
        ffi::requestScreenCaptureAccess()
    }
}

#[cfg(not(target_os = "macos"))]
mod imp {
    use super::{CapturePermissions, PermStatus};

    pub fn check() -> CapturePermissions {
        CapturePermissions {
            microphone: PermStatus::NotApplicable,
            screen_recording: PermStatus::NotApplicable,
        }
    }

    pub fn request_mic() -> PermStatus {
        PermStatus::NotApplicable
    }

    pub fn request_screen() -> bool {
        // No TCC gate on Linux/Windows — treat as "already allowed" so
        // callers don't surface a spurious denial chip.
        true
    }
}

/// Read the current grant state for both capture permissions. Cheap
/// status read — never blocks on a prompt — so it stays safe to call
/// on the hot Start path.
pub fn check_capture_permissions() -> CapturePermissions {
    imp::check()
}

/// Trigger the OS mic-permission prompt (macOS) and return the
/// resulting status. No-op `not_applicable` off macOS.
pub fn request_mic_permission() -> PermStatus {
    imp::request_mic()
}

/// Trigger the OS screen-recording prompt (macOS) and return whether
/// access is granted. Always `true` off macOS.
pub fn request_screen_capture_access() -> bool {
    imp::request_screen()
}

#[cfg(test)]
mod tests {
    use super::*;

    // On the Linux dev box / CI, both permissions must report
    // `not_applicable` and the request helpers must no-op cleanly.
    // The macOS arm is exercised only by the `build-agent.yml` build
    // leg (Linux `cargo test` cannot reach the `#[cfg(macos)]` code).
    #[cfg(not(target_os = "macos"))]
    #[test]
    fn non_mac_permissions_are_not_applicable() {
        let p = check_capture_permissions();
        assert_eq!(p.microphone, PermStatus::NotApplicable);
        assert_eq!(p.screen_recording, PermStatus::NotApplicable);
        assert_eq!(request_mic_permission(), PermStatus::NotApplicable);
        assert!(request_screen_capture_access());
    }
}
